plugins/base/src/main/kotlin/resolvers/external/ExternalLocationProviderFactory.kt

```diff
@@ -0,0 +1,24 @@
+package org.jetbrains.dokka.base.resolvers.external
+
+import org.jetbrains.dokka.links.DRI
+
+
+interface ExternalLocationProvider {
+
+    val param: String
+    fun DRI.toLocation(): String
+}
+
+interface ExternalLocationProviderFactory {
+    fun getExternalLocationProvider(param: String): ExternalLocationProvider?
+}
+
+abstract class ExternalLocationProviderFactoryWithCache : ExternalLocationProviderFactory {
+
+    private val locationProviders: MutableList<ExternalLocationProvider> = mutableListOf()
+
+    override fun getExternalLocationProvider(param: String): ExternalLocationProvider? =
```

Comment:
If someone overrides this method, whole logic is gonna collapse

Reply:
I was worried about the same thing, but didn't have a clue how to inherit from interface and keep the method closed. Check out my second attempt, which actually have the default implementation closed (I don't know however if this trick is allowed)

Reply:
Sorry, my bad, this is not an open class, no one is gonna override this :P You can safely leave it as it is

Reply:
You sure? Now it is not an open class, but before my updates it was abstract, somebody could override `getExternalLocationProvider`.